// @stacks/transactions/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSTXPostCondition = exports.createNonFungiblePostCondition = exports.createFungiblePostCondition = exports.serializePayload = exports.isSingleSig = exports.emptyMessageSignature = exports.BufferReader = exports.deserializeTransaction = exports.StacksTransaction = void 0;
var transaction_1 = require("./transaction");
Object.defineProperty(exports, "StacksTransaction", { enumerable: true, get: function () { return transaction_1.StacksTransaction; } });
Object.defineProperty(exports, "deserializeTransaction", { enumerable: true, get: function () { return transaction_1.deserializeTransaction; } });
var bufferReader_1 = require("./bufferReader");
Object.defineProperty(exports, "BufferReader", { enumerable: true, get: function () { return bufferReader_1.BufferReader; } });
var authorization_1 = require("./authorization");
Object.defineProperty(exports, "emptyMessageSignature", { enumerable: true, get: function () { return authorization_1.emptyMessageSignature; } });
Object.defineProperty(exports, "isSingleSig", { enumerable: true, get: function () { return authorization_1.isSingleSig; } });
var payload_1 = require("./payload");
Object.defineProperty(exports, "serializePayload", { enumerable: true, get: function () { return payload_1.serializePayload; } });
var postcondition_1 = require("./postcondition");
Object.defineProperty(exports, "createFungiblePostCondition", { enumerable: true, get: function () { return postcondition_1.createFungiblePostCondition; } });
Object.defineProperty(exports, "createNonFungiblePostCondition", { enumerable: true, get: function () { return postcondition_1.createNonFungiblePostCondition; } });
Object.defineProperty(exports, "createSTXPostCondition", { enumerable: true, get: function () { return postcondition_1.createSTXPostCondition; } });
__exportStar(require("./clarity"), exports);
__exportStar(require("./keys"), exports);
__exportStar(require("./builders"), exports);
__exportStar(require("./types"), exports);
__exportStar(require("./constants"), exports);
__exportStar(require("./contract-abi"), exports);
__exportStar(require("./signer"), exports);
__exportStar(require("./authorization"), exports);
__exportStar(require("./utils"), exports);
__exportStar(require("./common"), exports);
__exportStar(require("./signature"), exports);
__exportStar(require("./structuredDataSignature"), exports);
__exportStar(require("./postcondition-types"), exports);
