// @stacks/auth/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserSession = exports.verifyAuthRequestAndLoadManifest = exports.isRedirectUriValid = exports.isManifestUriValid = exports.doSignaturesMatchPublicKeys = exports.doPublicKeysMatchIssuer = exports.isIssuanceDateValid = exports.isExpirationDateValid = exports.verifyAuthResponse = exports.verifyAuthRequest = exports.fetchAppManifest = exports.getAuthRequestFromURL = exports.decryptPrivateKey = exports.makeAuthResponse = exports.makeAuthRequest = exports.AppConfig = void 0;
var appConfig_1 = require("./appConfig");
Object.defineProperty(exports, "AppConfig", { enumerable: true, get: function () { return appConfig_1.AppConfig; } });
var messages_1 = require("./messages");
Object.defineProperty(exports, "makeAuthRequest", { enumerable: true, get: function () { return messages_1.makeAuthRequest; } });
Object.defineProperty(exports, "makeAuthResponse", { enumerable: true, get: function () { return messages_1.makeAuthResponse; } });
Object.defineProperty(exports, "decryptPrivateKey", { enumerable: true, get: function () { return messages_1.decryptPrivateKey; } });
var provider_1 = require("./provider");
Object.defineProperty(exports, "getAuthRequestFromURL", { enumerable: true, get: function () { return provider_1.getAuthRequestFromURL; } });
Object.defineProperty(exports, "fetchAppManifest", { enumerable: true, get: function () { return provider_1.fetchAppManifest; } });
var verification_1 = require("./verification");
Object.defineProperty(exports, "verifyAuthRequest", { enumerable: true, get: function () { return verification_1.verifyAuthRequest; } });
Object.defineProperty(exports, "verifyAuthResponse", { enumerable: true, get: function () { return verification_1.verifyAuthResponse; } });
Object.defineProperty(exports, "isExpirationDateValid", { enumerable: true, get: function () { return verification_1.isExpirationDateValid; } });
Object.defineProperty(exports, "isIssuanceDateValid", { enumerable: true, get: function () { return verification_1.isIssuanceDateValid; } });
Object.defineProperty(exports, "doPublicKeysMatchIssuer", { enumerable: true, get: function () { return verification_1.doPublicKeysMatchIssuer; } });
Object.defineProperty(exports, "doSignaturesMatchPublicKeys", { enumerable: true, get: function () { return verification_1.doSignaturesMatchPublicKeys; } });
Object.defineProperty(exports, "isManifestUriValid", { enumerable: true, get: function () { return verification_1.isManifestUriValid; } });
Object.defineProperty(exports, "isRedirectUriValid", { enumerable: true, get: function () { return verification_1.isRedirectUriValid; } });
Object.defineProperty(exports, "verifyAuthRequestAndLoadManifest", { enumerable: true, get: function () { return verification_1.verifyAuthRequestAndLoadManifest; } });
__exportStar(require("./dids"), exports);
var userSession_1 = require("./userSession");
Object.defineProperty(exports, "UserSession", { enumerable: true, get: function () { return userSession_1.UserSession; } });
__exportStar(require("./constants"), exports);
__exportStar(require("./profile"), exports);
__exportStar(require("./userData"), exports);
