// @stacks/auth/dist/appConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppConfig = void 0;
const constants_1 = require("./constants");
const common_1 = require("@stacks/common");
class AppConfig {
    constructor(scopes, appDomain, redirectPath, manifestPath, coreNode, authenticatorURL) {
        var _a;
        if (scopes === void 0) { scopes = constants_1.DEFAULT_SCOPE.slice(); }
        if (appDomain === void 0) { appDomain = (_a = (0, common_1.getGlobalObject)('location', { returnEmptyObject: true })) === null || _a === void 0 ? void 0 : _a.origin; }
        if (redirectPath === void 0) { redirectPath = ''; }
        if (manifestPath === void 0) { manifestPath = '/manifest.json'; }
        if (coreNode === void 0) { coreNode = undefined; }
        if (authenticatorURL === void 0) { authenticatorURL = constants_1.DEFAULT_BLOCKSTACK_HOST; }
        this.appDomain = appDomain;
        this.scopes = scopes;
        this.redirectPath = redirectPath;
        this.manifestPath = manifestPath;
        this.coreNode = coreNode;
        this.authenticatorURL = authenticatorURL;
    }
    redirectURI() {
        return `${this.appDomain}${this.redirectPath}`;
    }
    manifestURI() {
        return `${this.appDomain}${this.manifestPath}`;
    }
}
exports.AppConfig = AppConfig;
