// @stacks/transactions/dist/clarity/serialize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeCV = void 0;
const common_1 = require("@stacks/common");
const types_1 = require("../types");
const postcondition_types_1 = require("../postcondition-types");
const constants_1 = require("./constants");
const utils_1 = require("../utils");
const errors_1 = require("../errors");
const constants_2 = require("../constants");
function bufferWithTypeID(typeId, buffer) {
    const id = common_1.Buffer.from([typeId]);
    return common_1.Buffer.concat([id, buffer]);
}
function serializeBoolCV(value) {
    return common_1.Buffer.from([value.type]);
}
function serializeOptionalCV(cv) {
    if (cv.type === constants_1.ClarityType.OptionalNone) {
        return common_1.Buffer.from([cv.type]);
    }
    else {
        return bufferWithTypeID(cv.type, serializeCV(cv.value));
    }
}
function serializeBufferCV(cv) {
    const length = common_1.Buffer.alloc(4);
    length.writeUInt32BE(cv.buffer.length, 0);
    return bufferWithTypeID(cv.type, common_1.Buffer.concat([length, cv.buffer]));
}
function serializeIntCV(cv) {
    const buffer = (0, common_1.toBuffer)((0, common_1.toTwos)(cv.value, BigInt(constants_2.CLARITY_INT_SIZE)), constants_2.CLARITY_INT_BYTE_SIZE);
    return bufferWithTypeID(cv.type, buffer);
}
function serializeUIntCV(cv) {
    const buffer = (0, common_1.toBuffer)(cv.value, constants_2.CLARITY_INT_BYTE_SIZE);
    return bufferWithTypeID(cv.type, buffer);
}
function serializeStandardPrincipalCV(cv) {
    return bufferWithTypeID(cv.type, (0, types_1.serializeAddress)(cv.address));
}
function serializeContractPrincipalCV(cv) {
    return bufferWithTypeID(cv.type, common_1.Buffer.concat([(0, types_1.serializeAddress)(cv.address), (0, types_1.serializeLPString)(cv.contractName)]));
}
function serializeResponseCV(cv) {
    return bufferWithTypeID(cv.type, serializeCV(cv.value));
}
function serializeListCV(cv) {
    const buffers = new utils_1.BufferArray();
    const length = common_1.Buffer.alloc(4);
    length.writeUInt32BE(cv.list.length, 0);
    buffers.push(length);
    for (const value of cv.list) {
        const serializedValue = serializeCV(value);
        buffers.push(serializedValue);
    }
    return bufferWithTypeID(cv.type, buffers.concatBuffer());
}
function serializeTupleCV(cv) {
    const buffers = new utils_1.BufferArray();
    const length = common_1.Buffer.alloc(4);
    length.writeUInt32BE(Object.keys(cv.data).length, 0);
    buffers.push(length);
    const lexicographicOrder = Object.keys(cv.data).sort((a, b) => {
        const bufA = common_1.Buffer.from(a);
        const bufB = common_1.Buffer.from(b);
        return bufA.compare(bufB);
    });
    for (const key of lexicographicOrder) {
        const nameWithLength = (0, postcondition_types_1.createLPString)(key);
        buffers.push((0, types_1.serializeLPString)(nameWithLength));
        const serializedValue = serializeCV(cv.data[key]);
        buffers.push(serializedValue);
    }
    return bufferWithTypeID(cv.type, buffers.concatBuffer());
}
function serializeStringCV(cv, encoding) {
    const buffers = new utils_1.BufferArray();
    const str = common_1.Buffer.from(cv.data, encoding);
    const len = common_1.Buffer.alloc(4);
    len.writeUInt32BE(str.length, 0);
    buffers.push(len);
    buffers.push(str);
    return bufferWithTypeID(cv.type, buffers.concatBuffer());
}
function serializeStringAsciiCV(cv) {
    return serializeStringCV(cv, 'ascii');
}
function serializeStringUtf8CV(cv) {
    return serializeStringCV(cv, 'utf8');
}
function serializeCV(value) {
    switch (value.type) {
        case constants_1.ClarityType.BoolTrue:
        case constants_1.ClarityType.BoolFalse:
            return serializeBoolCV(value);
        case constants_1.ClarityType.OptionalNone:
        case constants_1.ClarityType.OptionalSome:
            return serializeOptionalCV(value);
        case constants_1.ClarityType.Buffer:
            return serializeBufferCV(value);
        case constants_1.ClarityType.UInt:
            return serializeUIntCV(value);
        case constants_1.ClarityType.Int:
            return serializeIntCV(value);
        case constants_1.ClarityType.PrincipalStandard:
            return serializeStandardPrincipalCV(value);
        case constants_1.ClarityType.PrincipalContract:
            return serializeContractPrincipalCV(value);
        case constants_1.ClarityType.ResponseOk:
        case constants_1.ClarityType.ResponseErr:
            return serializeResponseCV(value);
        case constants_1.ClarityType.List:
            return serializeListCV(value);
        case constants_1.ClarityType.Tuple:
            return serializeTupleCV(value);
        case constants_1.ClarityType.StringASCII:
            return serializeStringAsciiCV(value);
        case constants_1.ClarityType.StringUTF8:
            return serializeStringUtf8CV(value);
        default:
            throw new errors_1.SerializationError('Unable to serialize. Invalid Clarity Value.');
    }
}
exports.serializeCV = serializeCV;
