// @stacks/encryption/dist/ec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyMessageSignatureRsv = exports.verifyMessageSignature = exports.verifyECDSA = exports.signECDSA = exports.decryptECIES = exports.encryptECIES = exports.eciesGetJsonStringLength = exports.getSignedCipherObjectWrapper = exports.getCipherObjectWrapper = exports.getBufferFromBN = exports.getHexFromBN = exports.aes256CbcEncrypt = exports.InvalidPublicKeyReason = void 0;
const hmac_1 = require("@noble/hashes/hmac");
const sha256_1 = require("@noble/hashes/sha256");
const secp256k1_1 = require("@noble/secp256k1");
const common_1 = require("@stacks/common");
const aesCipher_1 = require("./aesCipher");
const hmacSha256_1 = require("./hmacSha256");
const keys_1 = require("./keys");
const messageSignature_1 = require("./messageSignature");
const sha2Hash_1 = require("./sha2Hash");
const utils_1 = require("./utils");
secp256k1_1.utils.hmacSha256Sync = (key, ...msgs) => {
    const h = hmac_1.hmac.create(sha256_1.sha256, key);
    msgs.forEach(msg => h.update(msg));
    return h.digest();
};
var InvalidPublicKeyReason;
(function (InvalidPublicKeyReason) {
    InvalidPublicKeyReason["InvalidFormat"] = "InvalidFormat";
    InvalidPublicKeyReason["IsNotPoint"] = "IsNotPoint";
})(InvalidPublicKeyReason = exports.InvalidPublicKeyReason || (exports.InvalidPublicKeyReason = {}));
async function aes256CbcEncrypt(iv, key, plaintext) {
    const cipher = await (0, aesCipher_1.createCipher)();
    const result = await cipher.encrypt('aes-256-cbc', key, iv, plaintext);
    return result;
}
exports.aes256CbcEncrypt = aes256CbcEncrypt;
async function aes256CbcDecrypt(iv, key, ciphertext) {
    const cipher = await (0, aesCipher_1.createCipher)();
    const result = await cipher.decrypt('aes-256-cbc', key, iv, ciphertext);
    return result;
}
async function hmacSha256(key, content) {
    const hmacSha256 = await (0, hmacSha256_1.createHmacSha256)();
    return hmacSha256.digest(key, content);
}
function equalConstTime(b1, b2) {
    if (b1.length !== b2.length) {
        return false;
    }
    let res = 0;
    for (let i = 0; i < b1.length; i++) {
        res |= b1[i] ^ b2[i];
    }
    return res === 0;
}
function sharedSecretToKeys(sharedSecret) {
    const hashedSecret = (0, sha2Hash_1.hashSha512Sync)(sharedSecret);
    return {
        encryptionKey: hashedSecret.slice(0, 32),
        hmacKey: hashedSecret.slice(32),
    };
}
function allHexChars(maybe) {
    return maybe.match(/^[0-9a-f]+$/i) !== null;
}
function isValidPublicKey(pub) {
    const invalidFormat = {
        result: false,
        reason_data: 'Invalid public key format',
        reason: InvalidPublicKeyReason.InvalidFormat,
    };
    const invalidPoint = {
        result: false,
        reason_data: 'Public key is not a point',
        reason: InvalidPublicKeyReason.IsNotPoint,
    };
    if (pub.length !== 66 && pub.length !== 130)
        return invalidFormat;
    const firstByte = pub.slice(0, 2);
    if (pub.length === 130 && firstByte !== '04')
        return invalidFormat;
    if (pub.length === 66 && firstByte !== '02' && firstByte !== '03')
        return invalidFormat;
    if (!allHexChars(pub))
        return invalidFormat;
    try {
        const point = secp256k1_1.Point.fromHex(pub);
        point.assertValidity();
        return {
            result: true,
            reason_data: null,
            reason: null,
        };
    }
    catch (e) {
        return invalidPoint;
    }
}
function getHexFromBN(bnInput) {
    const hexOut = bnInput.toString(16);
    if (hexOut.length === 64) {
        return hexOut;
    }
    else if (hexOut.length < 64) {
        const padding = '0'.repeat(64 - hexOut.length);
        return `${padding}${hexOut}`;
    }
    else {
        throw new Error('Generated a > 32-byte BN for encryption. Failing.');
    }
}
exports.getHexFromBN = getHexFromBN;
function getBufferFromBN(bnInput) {
    const result = (0, common_1.toBuffer)(bnInput, 32);
    if (result.byteLength !== 32) {
        throw new Error('Failed to generate a 32-byte buffer instance');
    }
    return result;
}
exports.getBufferFromBN = getBufferFromBN;
function getCipherObjectWrapper(opts) {
    const shell = {
        iv: '',
        ephemeralPK: '',
        mac: '',
        cipherText: '',
        wasString: !!opts.wasString,
    };
    if (opts.cipherTextEncoding === 'base64') {
        shell.cipherTextEncoding = 'base64';
    }
    const ivLength = 32;
    const ephemeralPKLength = 66;
    const macLength = 64;
    return {
        payloadValuesLength: ivLength + ephemeralPKLength + macLength,
        payloadShell: JSON.stringify(shell),
    };
}
exports.getCipherObjectWrapper = getCipherObjectWrapper;
function getSignedCipherObjectWrapper(payloadShell) {
    const shell = {
        signature: '',
        publicKey: '',
        cipherText: payloadShell,
    };
    const signatureLength = 144;
    const publicKeyLength = 66;
    return {
        signedPayloadValuesLength: signatureLength + publicKeyLength,
        signedPayloadShell: JSON.stringify(shell),
    };
}
exports.getSignedCipherObjectWrapper = getSignedCipherObjectWrapper;
function eciesGetJsonStringLength(opts) {
    const { payloadShell, payloadValuesLength } = getCipherObjectWrapper(opts);
    const cipherTextLength = (0, utils_1.getAesCbcOutputLength)(opts.contentLength);
    let encodedCipherTextLength;
    if (!opts.cipherTextEncoding || opts.cipherTextEncoding === 'hex') {
        encodedCipherTextLength = cipherTextLength * 2;
    }
    else if (opts.cipherTextEncoding === 'base64') {
        encodedCipherTextLength = (0, utils_1.getBase64OutputLength)(cipherTextLength);
    }
    else {
        throw new Error(`Unexpected cipherTextEncoding "${opts.cipherTextEncoding}"`);
    }
    if (!opts.sign) {
        return payloadShell.length + payloadValuesLength + encodedCipherTextLength;
    }
    else {
        const { signedPayloadShell, signedPayloadValuesLength } = getSignedCipherObjectWrapper(payloadShell);
        return (signedPayloadShell.length +
            signedPayloadValuesLength +
            payloadValuesLength +
            encodedCipherTextLength);
    }
}
exports.eciesGetJsonStringLength = eciesGetJsonStringLength;
async function encryptECIES(publicKey, content, wasString, cipherTextEncoding) {
    const validity = isValidPublicKey(publicKey);
    if (!validity.result) {
        throw validity;
    }
    const ephemeralPrivateKey = secp256k1_1.utils.randomPrivateKey();
    const ephemeralPublicKey = (0, secp256k1_1.getPublicKey)(ephemeralPrivateKey, true);
    let sharedSecret = (0, secp256k1_1.getSharedSecret)(ephemeralPrivateKey, publicKey, true);
    sharedSecret = sharedSecret.slice(1);
    const sharedKeys = sharedSecretToKeys(common_1.Buffer.from(sharedSecret));
    const initializationVector = secp256k1_1.utils.randomBytes(16);
    const cipherText = await aes256CbcEncrypt(common_1.Buffer.from(initializationVector), sharedKeys.encryptionKey, content);
    const macData = (0, common_1.concatBytes)(initializationVector, ephemeralPublicKey, cipherText);
    const mac = await hmacSha256(sharedKeys.hmacKey, common_1.Buffer.from(macData));
    let cipherTextString;
    if (!cipherTextEncoding || cipherTextEncoding === 'hex') {
        cipherTextString = (0, common_1.bytesToHex)(cipherText);
    }
    else if (cipherTextEncoding === 'base64') {
        cipherTextString = cipherText.toString('base64');
    }
    else {
        throw new Error(`Unexpected cipherTextEncoding "${cipherTextEncoding}"`);
    }
    const result = {
        iv: (0, common_1.bytesToHex)(initializationVector),
        ephemeralPK: (0, common_1.bytesToHex)(ephemeralPublicKey),
        cipherText: cipherTextString,
        mac: (0, common_1.bytesToHex)(mac),
        wasString,
    };
    if (cipherTextEncoding && cipherTextEncoding !== 'hex') {
        result.cipherTextEncoding = cipherTextEncoding;
    }
    return result;
}
exports.encryptECIES = encryptECIES;
async function decryptECIES(privateKey, cipherObject) {
    if (!cipherObject.ephemeralPK) {
        throw new common_1.FailedDecryptionError('Unable to get public key from cipher object. ' +
            'You might be trying to decrypt an unencrypted object.');
    }
    const ephemeralPK = cipherObject.ephemeralPK;
    let sharedSecret = (0, secp256k1_1.getSharedSecret)(privateKey, ephemeralPK, true);
    sharedSecret = sharedSecret.slice(1);
    const sharedKeys = sharedSecretToKeys(common_1.Buffer.from(sharedSecret));
    const ivBuffer = (0, common_1.hexToBytes)(cipherObject.iv);
    let cipherTextBuffer;
    if (!cipherObject.cipherTextEncoding || cipherObject.cipherTextEncoding === 'hex') {
        cipherTextBuffer = common_1.Buffer.from(cipherObject.cipherText, 'hex');
    }
    else if (cipherObject.cipherTextEncoding === 'base64') {
        cipherTextBuffer = common_1.Buffer.from(cipherObject.cipherText, 'base64');
    }
    else {
        throw new Error(`Unexpected cipherTextEncoding "${cipherObject.cipherText}"`);
    }
    const macData = (0, common_1.concatBytes)(ivBuffer, (0, common_1.hexToBytes)(ephemeralPK), cipherTextBuffer);
    const actualMac = await hmacSha256(sharedKeys.hmacKey, common_1.Buffer.from(macData));
    const expectedMac = (0, common_1.hexToBytes)(cipherObject.mac);
    if (!equalConstTime(common_1.Buffer.from(expectedMac), actualMac)) {
        throw new common_1.FailedDecryptionError('Decryption failed: failure in MAC check');
    }
    const plainText = await aes256CbcDecrypt(common_1.Buffer.from(ivBuffer), sharedKeys.encryptionKey, cipherTextBuffer);
    if (cipherObject.wasString) {
        return plainText.toString();
    }
    else {
        return plainText;
    }
}
exports.decryptECIES = decryptECIES;
function signECDSA(privateKey, content) {
    const contentBuffer = content instanceof common_1.Buffer ? content : common_1.Buffer.from(content);
    const publicKey = (0, keys_1.getPublicKeyFromPrivate)(privateKey);
    const contentHash = (0, sha2Hash_1.hashSha256Sync)(contentBuffer);
    const signature = (0, secp256k1_1.signSync)(contentHash, privateKey);
    return {
        signature: (0, common_1.bytesToHex)(signature),
        publicKey,
    };
}
exports.signECDSA = signECDSA;
function getBuffer(content) {
    if (content instanceof common_1.Buffer)
        return content;
    else if (content instanceof ArrayBuffer)
        return common_1.Buffer.from(content);
    else
        return common_1.Buffer.from(content);
}
function verifyECDSA(content, publicKey, signature) {
    const contentBuffer = getBuffer(content);
    const contentHash = (0, sha2Hash_1.hashSha256Sync)(contentBuffer);
    return (0, secp256k1_1.verify)(signature, contentHash, publicKey, { strict: false });
}
exports.verifyECDSA = verifyECDSA;
function verifyMessageSignature({ signature, message, publicKey, }) {
    const { r, s } = (0, common_1.parseRecoverableSignatureVrs)(signature);
    const sig = new secp256k1_1.Signature((0, common_1.hexToBigInt)(r), (0, common_1.hexToBigInt)(s));
    const hashedMsg = typeof message === 'string' ? (0, messageSignature_1.hashMessage)(message) : message;
    return (0, secp256k1_1.verify)(sig, hashedMsg, publicKey, { strict: false });
}
exports.verifyMessageSignature = verifyMessageSignature;
function verifyMessageSignatureRsv({ signature, message, publicKey, }) {
    return verifyMessageSignature({
        signature: (0, common_1.signatureRsvToVrs)(signature),
        message,
        publicKey,
    });
}
exports.verifyMessageSignatureRsv = verifyMessageSignatureRsv;
